File: server/src/github/github-stats-service.ts
import { findGitHubStatsCache, getActivityYearsSetting, upsertGitHubStatsCache } from "../db.js";
import { getGithubUsername } from "../config.js";
import { longestStreak, peakDay, type ActivityDay } from "./github-stats-calculator.js";
import {
  activityRangeForYear,
  flattenContributionWeeks,
  mapContributionWeeks,
  trimContributionWeeks,
  type ContributionWeek,
} from "./contribution-calendar.js";
import type { GitHubActivity, GitHubActivityYears } from "./types.js";

const CACHE_TTL_MS = 6 * 60 * 60 * 1000;
const LIFETIME_START_YEAR = 2008;
const USER_AGENT = "portfolio-app";
const ACTIVITY_CACHE_VERSION = "v2";

type CollectionNode = {
  contributionCalendar: {
    totalContributions: number;
    weeks: ContributionWeek[];
  };
  commitContributionsByRepository: Array<{
    repository: { nameWithOwner: string; url: string; isPrivate: boolean };
    contributions: { totalCount: number };
  }>;
};

async function fetchCollection(
  accessToken: string,
  username: string,
  from: string,
  to: string
): Promise<CollectionNode> {
  const query = `
    query($login: String!, $from: DateTime!, $to: DateTime!) {
      user(login: $login) {
        contributionsCollection(from: $from, to: $to) {
          contributionCalendar {
            totalContributions
            weeks {
              contributionDays {
                date
                contributionCount
              }
            }
          }
          commitContributionsByRepository(maxRepositories: 20) {
            repository {
              nameWithOwner
              url
              isPrivate
            }
            contributions {
              totalCount
            }
          }
        }
      }
    }
  `;

  const response = await fetch("https://api.github.com/graphql", {
    method: "POST",
    headers: {
      Authorization: `Bearer ${accessToken}`,
      "Content-Type": "application/json",
      "User-Agent": USER_AGENT,
    },
    body: JSON.stringify({
      query,
      variables: { login: username, from, to },
    }),
  });

  if (!response.ok) {
    const body = await response.text();
    throw new Error(`Failed to fetch GitHub activity (HTTP ${response.status}): ${body}`);
  }

  const decoded = (await response.json()) as {
    data?: { user?: { contributionsCollection?: CollectionNode } };
    errors?: Array<{ message: string }>;
  };

  if (decoded.errors?.length && !decoded.data?.user) {
    throw new Error(`GitHub GraphQL error: ${decoded.errors.map((e) => e.message).join("; ")}`);
  }

  const collection = decoded.data?.user?.contributionsCollection;
  if (!collection) {
    throw Object.assign(new Error(`GitHub user '${username}' not found or has no public activity`), { status: 404 });
  }

  return collection;
}

function buildActivityPayload(year: number, collection: CollectionNode): GitHubActivity {
  const trimmedWeeks = trimContributionWeeks(collection.contributionCalendar.weeks, year);
  const days = flattenContributionWeeks(trimmedWeeks);
  const peak = peakDay(days);
  const topPublicRepos = collection.commitContributionsByRepository
    .filter((item) => !item.repository.isPrivate)
    .sort((a, b) => b.contributions.totalCount - a.contributions.totalCount)
    .slice(0, 8)
    .map((item) => ({
      fullName: item.repository.nameWithOwner,
      url: item.repository.url,
      commits: item.contributions.totalCount,
    }));

  return {
    year,
    days,
    weeks: mapContributionWeeks(trimmedWeeks),
    stats: {
      lifetimeCommits: 0,
      peakCommitsInDay: peak.count,
      peakDay: peak.date,
      longestStreak: longestStreak(days),
      yearCommits: collection.contributionCalendar.totalContributions,
    },
    topPublicRepos,
  };
}

async function fetchLifetimeStats(accessToken: string, username: string) {
  const cacheKey = `activity:${username}:lifetime`;
  const now = Date.now();
  const cached = await findGitHubStatsCache(cacheKey);
  if (cached && cached.expires_at > now) {
    return JSON.parse(cached.payload_json) as GitHubActivity["stats"];
  }

  const currentYear = new Date().getFullYear();
  const allDays: ActivityDay[] = [];
  let lifetimeCommits = 0;

  for (let year = LIFETIME_START_YEAR; year <= currentYear; year++) {
    const { from, to } = activityRangeForYear(year);
    try {
      const collection = await fetchCollection(accessToken, username, from, to);
      lifetimeCommits += collection.contributionCalendar.totalContributions;
      const trimmedWeeks = trimContributionWeeks(collection.contributionCalendar.weeks, year);
      allDays.push(...flattenContributionWeeks(trimmedWeeks));
    } catch (err) {
      if ((err as { status?: number }).status === 404) continue;
      throw err;
    }
  }

  const peak = peakDay(allDays);
  const stats = {
    lifetimeCommits,
    peakCommitsInDay: peak.count,
    peakDay: peak.date,
    longestStreak: longestStreak(allDays),
    yearCommits: 0,
  };

  await upsertGitHubStatsCache(cacheKey, JSON.stringify(stats), now + CACHE_TTL_MS);
  return stats;
}

function defaultActivityYears(): number[] {
  const currentYear = new Date().getFullYear();
  return Array.from({ length: 6 }, (_, index) => currentYear - index);
}

function normalizeActivityYears(years: number[]): number[] {
  const currentYear = new Date().getFullYear();
  return [...new Set(years)]
    .filter((year) => Number.isInteger(year) && year >= LIFETIME_START_YEAR && year <= currentYear)
    .sort((a, b) => b - a);
}

export async function resolveActivityYears(): Promise<number[]> {
  const configured = await getActivityYearsSetting();
  if (!configured) return defaultActivityYears();
  const normalized = normalizeActivityYears(configured);
  return normalized.length > 0 ? normalized : defaultActivityYears();
}

export async function getActivityYears(accessToken: string): Promise<GitHubActivityYears> {
  const username = getGithubUsername();
  const years = await resolveActivityYears();
  return { years, username };
}

export async function getActivity(accessToken: string, year: number): Promise<GitHubActivity> {
  const username = getGithubUsername();
  const cacheKey = `activity:${username}:${year}:${ACTIVITY_CACHE_VERSION}`;
  const now = Date.now();
  const cached = await findGitHubStatsCache(cacheKey);

  if (cached && cached.expires_at > now) {
    return JSON.parse(cached.payload_json) as GitHubActivity;
  }

  const { from, to } = activityRangeForYear(year);
  const collection = await fetchCollection(accessToken, username, from, to);
  const yearPayload = buildActivityPayload(year, collection);
  const lifetime = await fetchLifetimeStats(accessToken, username);

  const payload: GitHubActivity = {
    ...yearPayload,
    stats: {
      ...lifetime,
      yearCommits: yearPayload.stats.yearCommits,
    },
  };

  await upsertGitHubStatsCache(cacheKey, JSON.stringify(payload), now + CACHE_TTL_MS);
  return payload;
}

File: server/Sources/App/Services/GitHubStatsService.swift
import Foundation
import Vapor

struct GitHubStatsService {
    let client: Client
    let cacheRepository: GitHubStatsCacheRepository
    let siteSettingRepository: SiteSettingRepository
    let accessToken: String
    let username: String

    private let cacheTTL = 6 * 60 * 60 * 1000
    private let activityCacheVersion = "v2"

    func availableYears() async throws -> GitHubActivityYearsDTO {
        let years = try await siteSettingRepository.resolveActivityYears()
        return GitHubActivityYearsDTO(years: years, username: username)
    }

    func activity(for year: Int) async throws -> GitHubActivityDTO {
        let cacheKey = "activity:\(username):\(year):\(activityCacheVersion)"
        let now = nowMillis()

        if
            let cached = try await cacheRepository.find(key: cacheKey),
            cached.expiresAt > now,
            let data = cached.payloadJson.data(using: .utf8),
            let payload = try? JSONDecoder().decode(GitHubActivityDTO.self, from: data)
        {
            return payload
        }

        let range = ContributionCalendar.activityRange(for: year)
        let yearPayload = try await fetchActivity(year: year, from: range.from, to: range.to)
        let lifetimeStats = try await fetchLifetimeStats()
        let payload = mergeLifetime(into: yearPayload, lifetime: lifetimeStats)

        if let encoded = try? JSONEncoder().encode(payload), let json = String(data: encoded, encoding: .utf8) {
            try await cacheRepository.upsert(
                key: cacheKey,
                payloadJson: json,
                expiresAt: now + cacheTTL
            )
        }

        return payload
    }

    private func fetchActivity(year: Int, from: String, to: String) async throws -> GitHubActivityDTO {
        let collection = try await fetchCollection(from: from, to: to)
        let trimmedWeeks = trimWeeks(collection.contributionCalendar.weeks, year: year)
        let days = ContributionCalendar.flattenDays(from: trimmedWeeks)
        let weeks = ContributionCalendar.mapWeeks(trimmedWeeks)

        let calculatorDays = days.map { GitHubStatsCalculator.Day(date: $0.date, count: $0.count) }
        let peak = GitHubStatsCalculator.peakDay(from: calculatorDays)
        let stats = GitHubActivityStatsDTO(
            lifetimeCommits: 0,
            peakCommitsInDay: peak.count,
            peakDay: peak.date,
            longestStreak: GitHubStatsCalculator.longestStreak(from: calculatorDays),
            yearCommits: collection.contributionCalendar.totalContributions
        )

        let topPublicRepos = collection.commitContributionsByRepository
            .filter { !$0.repository.isPrivate }
            .sorted { $0.contributions.totalCount > $1.contributions.totalCount }
            .prefix(8)
            .map {
                GitHubTopRepoDTO(
                    fullName: $0.repository.nameWithOwner,
                    url: $0.repository.url,
                    commits: $0.contributions.totalCount
                )
            }

        return GitHubActivityDTO(
            year: year,
            days: days,
            weeks: weeks,
            stats: stats,
            topPublicRepos: Array(topPublicRepos)
        )
    }

    private func fetchLifetimeStats() async throws -> GitHubActivityStatsDTO {
        let cacheKey = "activity:\(username):lifetime"
        let now = nowMillis()

        if
            let cached = try await cacheRepository.find(key: cacheKey),
            cached.expiresAt > now,
            let data = cached.payloadJson.data(using: .utf8),
            let payload = try? JSONDecoder().decode(GitHubActivityStatsDTO.self, from: data)
        {
            return payload
        }

        let currentYear = Calendar.current.component(.year, from: Date())
        var allDays: [GitHubStatsCalculator.Day] = []
        var lifetimeCommits = 0

        for year in ActivityYearsSettings.lifetimeStartYear...currentYear {
            let range = ContributionCalendar.activityRange(for: year)

            guard let collection = try await fetchCollectionIfAvailable(from: range.from, to: range.to) else {
                continue
            }

            lifetimeCommits += collection.contributionCalendar.totalContributions
            let trimmedWeeks = trimWeeks(collection.contributionCalendar.weeks, year: year)
            let days = ContributionCalendar.flattenDays(from: trimmedWeeks)
            allDays.append(contentsOf: days.map { GitHubStatsCalculator.Day(date: $0.date, count: $0.count) })
        }

        let peak = GitHubStatsCalculator.peakDay(from: allDays)
        let stats = GitHubActivityStatsDTO(
            lifetimeCommits: lifetimeCommits,
            peakCommitsInDay: peak.count,
            peakDay: peak.date,
            longestStreak: GitHubStatsCalculator.longestStreak(from: allDays),
            yearCommits: 0
        )

        if let encoded = try? JSONEncoder().encode(stats), let json = String(data: encoded, encoding: .utf8) {
            try await cacheRepository.upsert(
                key: cacheKey,
                payloadJson: json,
                expiresAt: now + cacheTTL
            )
        }

        return stats
    }

    private func mergeLifetime(
        into yearPayload: GitHubActivityDTO,
        lifetime: GitHubActivityStatsDTO
    ) -> GitHubActivityDTO {
        GitHubActivityDTO(
            year: yearPayload.year,
            days: yearPayload.days,
            weeks: yearPayload.weeks,
            stats: GitHubActivityStatsDTO(
                lifetimeCommits: lifetime.lifetimeCommits,
                peakCommitsInDay: lifetime.peakCommitsInDay,
                peakDay: lifetime.peakDay,
                longestStreak: lifetime.longestStreak,
                yearCommits: yearPayload.stats.yearCommits
            ),
            topPublicRepos: yearPayload.topPublicRepos
        )
    }

    private func trimWeeks(
        _ weeks: [GraphQLResponse.WeekNode],
        year: Int
    ) -> [ContributionCalendar.Week] {
        let mapped = weeks.map { week in
            ContributionCalendar.Week(contributionDays: week.contributionDays.map {
                ContributionCalendar.Week.Day(date: $0.date, contributionCount: $0.contributionCount)
            })
        }
        return ContributionCalendar.trimWeeks(mapped, year: year)
    }

    private func fetchCollectionIfAvailable(from: String, to: String) async throws -> GraphQLResponse.CollectionNode? {
        do {
            return try await fetchCollection(from: from, to: to)
        } catch let error as ServiceError {
            if case .notFound = error {
                return nil
            }
            throw error
        }
    }

    private func fetchCollection(from: String, to: String) async throws -> GraphQLResponse.CollectionNode {
        let query = """
        query($login: String!, $from: DateTime!, $to: DateTime!) {
          user(login: $login) {
            contributionsCollection(from: $from, to: $to) {
              contributionCalendar {
                totalContributions
                weeks {
                  contributionDays {
                    date
                    contributionCount
                  }
                }
              }
              commitContributionsByRepository(maxRepositories: 20) {
                repository {
                  nameWithOwner
                  url
                  isPrivate
                }
                contributions {
                  totalCount
                }
              }
            }
          }
        }
        """

        let response = try await client.post("https://api.github.com/graphql") { request in
            request.headers.bearerAuthorization = BearerAuthorization(token: accessToken)
            request.headers.replaceOrAdd(name: .userAgent, value: "portfolio-app")
            request.headers.contentType = .json
            try request.content.encode(GraphQLRequest(
                query: query,
                variables: [
                    "login": username,
                    "from": from,
                    "to": to
                ]
            ))
        }

        guard response.status == .ok else {
            let body = response.body.map { String(buffer: $0) } ?? ""
            throw ServiceError.badRequest("Failed to fetch GitHub activity (HTTP \(response.status.code)): \(body)")
        }

        let decoded = try response.content.decode(GraphQLResponse.self)

        if let errors = decoded.errors, !errors.isEmpty {
            let messages = errors.map(\.message).joined(separator: "; ")
            if decoded.data?.user == nil {
                throw ServiceError.badRequest("GitHub GraphQL error: \(messages)")
            }
        }

        guard let collection = decoded.data?.user?.contributionsCollection else {
            throw ServiceError.notFound("GitHub user '\(username)' not found or has no public activity")
        }

        return collection
    }

    private struct GraphQLRequest: Content {
        let query: String
        let variables: [String: String]
    }

    private struct GraphQLResponse: Content {
        let data: DataNode?
        let errors: [GraphQLErrorNode]?

        struct DataNode: Content {
            let user: UserNode?
        }

        struct GraphQLErrorNode: Content {
            let message: String
        }

        struct UserNode: Content {
            let contributionsCollection: CollectionNode?
        }

        struct CollectionNode: Content {
            let contributionCalendar: CalendarNode
            let commitContributionsByRepository: [RepoContributionNode]
        }

        struct CalendarNode: Content {
            let totalContributions: Int
            let weeks: [WeekNode]
        }

        struct WeekNode: Content {
            let contributionDays: [DayNode]
        }

        struct DayNode: Content {
            let date: String
            let contributionCount: Int
        }

        struct RepoContributionNode: Content {
            let repository: RepositoryNode
            let contributions: CountNode
        }

        struct RepositoryNode: Content {
            let nameWithOwner: String
            let url: String
            let isPrivate: Bool
        }

        struct CountNode: Content {
            let totalCount: Int
        }
    }
}